### PR TITLE
modernize service worker example

### PR DESCRIPTION
**Pull request by SimonSiefke**
Tries to modernize the service worker example by
- replacing `Promise.then` with `async` / `await`
- replacing `XMLHttpRequest` with `fetch`
- replacing `var` with `const`

Not sure if this is a bit too over-engineered at some places, however in my testing, everything still worked as expected.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,74 +1,57 @@
...
+const createGalleryFigure = async (galleryImage) => {
+  const imageBlob = await getImageBlob(galleryImage.url);
+  const myImage = document.createElement("img");
+  const myCaption = document.createElement("caption");
+  const myFigure = document.createElement("figure");
+  myCaption.innerHTML =
+    "<strong>" +
+    galleryImage.name +
+    "</strong>: Taken by " +
+    galleryImage.credit;
```
Comment (wbamberg) [inline app.js]: While we are here could we get rid of this `innerHTML`? I think it would be OK to skip the `<strong>` for the name, like:

```suggestion
  myCaption.textContent = `${galleryImage.name}: Taken by ${galleryImage.credit}`;
```

...but if we do want to keep it we could of course create a `<strong>` element, as we do for the other elements here. I'm 50/50 on which to choose here :)

File: app.js [hunk 2]
```diff
@@ -1,74 +1,57 @@
-// register service worker
+import { Gallery } from "./image-list.js";
```
Comment (wbamberg) [inline app.js]: It's pretty much our convention in MDN to use single quotes, so while you are rewriting it might be worth changing double quotes to single quotes throughout? If you do though please do it in a separate commit to help me review.

File: app.js [hunk 3]
```diff
@@ -1,74 +1,57 @@
...
+  myImage.src = window.URL.createObjectURL(imageBlob);
+  myImage.setAttribute("alt", galleryImage.alt);
+  myFigure.append(myImage, myCaption);
+  imgSection.append(myFigure);
+};
 
-    }, function(Error) {
-      console.log(Error);
-    });
+window.onload = async () => {
```
Comment (wbamberg) [inline app.js]: Maybe rather than using `onload` we could move the `<script>` tag into the header and give it a `defer` attribute, then we guarantee this will run at the right time without needing an event. I think that would work, wouldn't it?

File: app.js [hunk 4]
```diff
@@ -1,74 +1,57 @@
...
+  if ("serviceWorker" in navigator) {
+    try {
+      await registerServiceWorker();
+    } catch (error) {
+      console.log("Registration failed with " + error);
+      return;
+    }
   }
+  await Promise.allSettled(Gallery.images.map(createGalleryFigure));
```
Comment (wbamberg) [inline app.js]: As I understand things, `createGalleryFigure` should not have to wait for `registerServiceWorker` to finish, so it seems wrong to use `await registerServiceWorker();` here - we're forcing them to be serialized when they don't need to be.

Similarly I don't think we need to wait for `Gallery.images.map(createGalleryFigure)`, because we don't have anything depending on its result. So I wonder if we could change `registerServiceWorker()` to be something like:

```js
const registerServiceWorker = async () => {
  try {
    const registration = await navigator.serviceWorker.register(
      "/sw-test/sw.js",
      {
        scope: "/sw-test/",
      }
    );
    if (registration.installing) {
      console.log("Service worker installing");
    } else if (registration.waiting) {
      console.log("Service worker installed");
    } else if (registration.active) {
      console.log("Service worker active");
    }
  } catch (error) {
    console.error(`Registration failed with ${error}`);
  }
};
```

...then change the top level to just:

```js
registerServiceWorker();
Gallery.images.map(createGalleryFigure);
```

Would that work?

File: app.js [hunk 5]
```diff
@@ -1,74 +1,57 @@
...
+const imgSection = document.querySelector("section");
 
-      myImage.src = imageURL;
-      myImage.setAttribute('alt', arrayResponse[1].alt);
-      myCaption.innerHTML = '<strong>' + arrayResponse[1].name + '</strong>: Taken by ' + arrayResponse[1].credit;
+const getImageBlob = async (url) => {
+  const imageResponse = await fetch(url);
+  if (!imageResponse.ok) {
+    throw new Error(
+      "Image didn't load successfully; error code:" + imageResponse.statusText
```
Comment (wbamberg) [inline app.js]:
```suggestion
      `Image didn't load successfully; error code: ${imageResponse.statusText}`
```

File: sw.js [hunk 6]
```diff
@@ -1,41 +1,62 @@
-self.addEventListener('install', function(event) {
+const addResourcesToCache = async (resources) => {
+  const cache = await caches.open("v1");
+  await cache.addAll(resources);
```
Comment (wbamberg) [inline sw.js:3]: I don't think we need `await` here because nothing is waiting for this.

File: sw.js [hunk 7]
```diff
@@ -1,41 +1,62 @@
...
+};
+
+const putInCache = async (request, response) => {
+  const cache = await caches.open("v1");
+  await cache.put(request, response);
+};
```
Comment (wbamberg) [inline sw.js]: In this case I would not have this as a separate function. It's only 2 lines, and is only called from one place, that's already async.

File: sw.js [hunk 8]
```diff
@@ -1,41 +1,62 @@
...
+
+const cacheFirst = async ({ request, fallbackUrl }) => {
```
Comment (wbamberg) [inline sw.js:11]: For this function is would be helpful to have a single line comment explaining the case which each bit is handling, like:

```js
// First try to get the resource from the cache
...
// Next try to get the resource from the network
...
```
etc

File: sw.js [hunk 9]
```diff
@@ -1,41 +1,62 @@
...
+    return new Response("Network error happened", {
+      status: 408,
+      headers: { "Content-Type": "text/plain" },
+    });
+  }
+  // response may be used only once
+  // we need to save clone to put one copy in cache
+  // and serve second one
+  putInCache(request, responseFromNetwork.clone());
+  return responseFromNetwork;
```
Comment (wbamberg) [inline sw.js]: As far as I can tell we only get to this bit if the `try` block succeeded (all the paths in the `catch` block return) so why not put this in the `try` block? Then each case is self-contained, and we could also move `let responseFromNetwork;` in there, and make it `const`.

File: sw.js [hunk 10]
```diff
@@ -1,41 +1,62 @@
...
+  }
+  let responseFromNetwork;
+  try {
+    responseFromNetwork = await fetch(request);
+  } catch (error) {
+    const fallbackResponse = await caches.match(fallbackUrl);
+    if (fallbackResponse) {
+      return fallbackResponse;
+    }
+    // when the even fallback response is not available,
```
Comment (wbamberg) [inline sw.js]:
```suggestion
    // when even the fallback response is not available,
```

File: app.js (hunk 5 above)
Comment (wbamberg) [inline app.js]: If this does throw (either `fetch` or the `throw` here) it's not caught anywhere. What would happen and is it what we want?

Maybe if any `getImageBlob` fails we should skip that one and move on to the next, so we should perhaps have a `try...catch` in `createGalleryFigure`, where in `catch` we just log an error. What do you think?

**Review by wbamberg — CHANGES_REQUESTED** [review]
Thanks for this PR, @SimonSiefke !

It looks really great to me, much more readable. I had some comments - some of them are not related to async/await, but while we are rewriting it they seem worth having.

Some of them I might be wrong about, happy to be corrected if so.

File: sw.js (hunk 7 above)
Comment (SimonSiefke) [reply sw.js]: Hm, one option would be to use await twice (which would block the return of responseFromNetwork)

```javascript
const cache = await caches.open('v1');
await cache.put(request, responseFromNetwork.clone()));
return responseFromNetwork;
```

Or just having one await (which would still block the return of responseFromNetwork):

```javascript
const cache = await caches.open('v1');
cache.put(request, responseFromNetwork.clone()));
return responseFromNetwork;
```

Another other option would be to use Promise.then (which I tried to avoid)
```javascript
caches
  .open('v1')
  .then((cache) => cache.put(request, responseFromNetwork.clone()));
return responseFromNetwork;
```

What do you think?

File: app.js (hunk 1 above)
Comment (SimonSiefke) [reply app.js]: I like textContent, done!

File: app.js (hunk 2 above)
Comment (SimonSiefke) [reply app.js]: done!

File: app.js (hunk 4 above)
Comment (SimonSiefke) [reply app.js]: That's great, I like it. Done!

File: app.js (hunk 3 above)
Comment (SimonSiefke) [reply app.js]: `<script type="module">` is always deferred, so defer attribute is not needed, i think. But I haved removed the onload listener. Done!

File: app.js (hunk 5 above)
Comment (SimonSiefke) [reply app.js]: Done!

File: sw.js (hunk 8 above)
Comment (SimonSiefke) [reply sw.js:11]: Added the comments!

File: sw.js (hunk 10 above)
Comment (SimonSiefke) [reply sw.js]: Fixed!

File: sw.js (hunk 9 above)
Comment (SimonSiefke) [reply sw.js]: Done!

File: app.js (hunk 5 above)
Comment (SimonSiefke) [reply app.js]: Oops, that would result in `Uncaught (in promise) Error: Image didn't load successfully` 

I've added the try-catch, just logging the error.

File: sw.js (hunk 6 above)
Comment (SimonSiefke) [reply sw.js:3]: Don't we want to wait until all resources have been added to the cache, when the service worker is installed? 

This is how it was before:

```javascript
self.addEventListener('install', function(event) {
  event.waitUntil(
    caches.open('v1').then(function(cache) {
      return cache.addAll([
        '/sw-test/',
        '/sw-test/index.html',
        '/sw-test/style.css',
        '/sw-test/app.js',
        '/sw-test/image-list.js',
        '/sw-test/star-wars-logo.jpg',
        '/sw-test/gallery/bountyHunters.jpg',
        '/sw-test/gallery/myLittleVader.jpg',
        '/sw-test/gallery/snowTroopers.jpg'
      ]);
    })
  );
});
```

Maybe I'm missing something...

File: sw.js (hunk 7 above)
Comment (wbamberg) [reply sw.js]: Oh I see. Yes, now you've explained it like that I like the original separate-function version :). But I still think you don't need the `await` in line 8, do you? Because nothing is waiting for it.

File: sw.js (hunk 6 above)
Comment (wbamberg) [reply sw.js:3]: Maybe I am :), but I think:

* `addResourcesToCache` is going to return a promise, anyway, because it's async
* `waitUntil` will wait until that returned promise is settled
* `await cache.addAll(resources)` won't affect that - it would only affect code inside `addResourcesToCache`. But there isn't any, we return right after that line anyway.

Put another way: AIUI, `await` inside `addResourcesToCache` isn't going to affect the code that called `addResourcesToCache` - either way, `addResourcesToCache` will return right away with a promise that will be settled later on, when the async stuff in `addResourcesToCache` has settled.

> Don't we want to wait until all resources have been added to the cache, when the service worker is installed?

What matter for this is that we are calling `waitUntil` on the promise that `addResourcesToCache` returns.

Does that make sense?

Comment (SimonSiefke) [reply sw.js:3]: > await cache.addAll(resources) won't affect that - it would only affect code inside addResourcesToCache. But there isn't any, we return right after that line anyway.

When calling (awaiting) this version of addResourcesToCache, the promise from addResourcesToCache is going to be resolved when cache.addAll(resources) is still in pending state.

```js
const addResourcesToCache = async (resources) => {
  const cache = await caches.open("v1");
  cache.addAll(resources);
}
```

For example, this

```js
const cache = {
  async addAll(resources) {
    await new Promise((resolve) => setTimeout(resolve, 0))
    console.log('resources have been added to cache')
  },
}

const addResourcesToCache = async (resources) => {
  cache.addAll(resources)
}

const main = async () => {
  await addResourcesToCache([])
  console.log('addResourcesToCache has finished')
}

main()

// prints out:
// addResourcesToCache has finished
// resources have been added to cache
```


Another problem would be error-handling with try-catch:


```js
const cache = {
  async addAll(resources) {
    await new Promise((resolve) => setTimeout(resolve, 0))
    throw new Error('oops')
  },
}

const addResourcesToCache = async (resources) => {
  cache.addAll(resources)
}

const main = async () => {
  try {
    await addResourcesToCache([])
  } catch (error) {
    console.log('error handling here')
  }
}

main()


// prints:
// Uncaught (in promise) Error: oops
```

Thoughts? :)

Comment (wbamberg) [reply sw.js:3]: OK I am convinced!

**Review by wbamberg — APPROVED** [review]
Thank you @SimonSiefke , this is looking great.
